netann/host_ann.go

```diff
@@ -0,0 +1,133 @@
+package netann
+
+import (
+	"net"
+	"sync"
+
+	"github.com/lightningnetwork/lnd/ticker"
+)
+
+// HostAnnouncerConfig is the main config for the HostAnnouncer.
+type HostAnnouncerConfig struct {
+	// Hosts is the set of hosts we should watch for IP changes.
+	Hosts []string
+
+	// RefreshTicker ticks each time we should check for any address
+	// changes.
+	RefreshTicker ticker.Ticker
+
+	// LookupHost performs DNS resolution on a given host and returns its
+	// addresses.
+	LookupHost func(string) (net.Addr, error)
+
+	// AnnounceNewIPs announces a new set of IP addresses for the backing
+	// Lightning node. The first set of addresses is the new set of
+	// addresses that we should advertise, while the other set are the
+	// stale addresses that we should no longer advertise.
+	AnnounceNewIPs func([]net.Addr, map[string]net.Addr) error
+}
+
+// HostAnnouncer is a sub-system that allows a user to specify a set of hosts
+// for lnd that will be continually resolved to notice any IP address changes.
+// If the target IP address for a host changes, then we'll generate a new
+// NodeAnnouncement that includes these new IPs.
+type HostAnnouncer struct {
+	cfg HostAnnouncerConfig
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+}
+
+// NewHostAnnouncer returns a new instance of the HostAnnouncer.
+func NewHostAnnouncer(cfg HostAnnouncerConfig) *HostAnnouncer {
+	return &HostAnnouncer{
+		cfg:  cfg,
+		quit: make(chan struct{}),
+	}
+}
+
+// Start starts the HostAnnouncer.
+func (h *HostAnnouncer) Start() error {
+	h.wg.Add(1)
+	go h.hostWatcher()
+
+	return nil
+}
+
+// Stop signals the HostAnnouncer for a graceful stop.
+func (h *HostAnnouncer) Stop() error {
+	close(h.quit)
+	h.wg.Wait()
+
+	return nil
+}
+
+// hostWatcher periodically attempts to resolve the IP for each host, updating
+// them if they change within the interval.
+func (h *HostAnnouncer) hostWatcher() {
+	defer h.wg.Done()
+
+	ipMapping := make(map[string]net.Addr)
+	refreshHosts := func() {
+
+		// We'll now run through each of our hosts to check if they had
+		// their backing IPs changed. If so, we'll want to re-announce
+		// them.
+		var addrsToUpdate []net.Addr
+		addrsToRemove := make(map[string]net.Addr)
+		for _, host := range h.cfg.Hosts {
+			newAddr, err := h.cfg.LookupHost(host)
+			if err != nil {
+				log.Warnf("unable to resolve IP for "+
+					"host %v: %v", host, err)
+				continue
+			}
+
+			oldAddr, ok := ipMapping[host]
+			if ok && oldAddr.String() == newAddr.String() {
```

Comment:
Maybe we should pass in the latest addresses we advertised? Otherwise, if we have `externalhosts` set, we'll always broadcast a node announcement on startup.

Reply:
What's wrong with broadcasting on start up?

Reply:
We used to moved away from it. We already have node announcement rebroadcast at a separate level, introducing that here as well seems like additional unnecessary complexity.

Reply:
Pushed up a diff to fix the log issue above, and also skip the logic to add the new address if this is the first time around. Updated the tests to remove the expectation that the addrs are announced the first time things start up....so things will just be delayed by 10 minutes.

Reply:
Sorry if I wasn't clear enough, I believe the previous behavior was correct. What I meant by my first comment was that we should pass in any existing node announcement addresses we have stored/configured to the `HostAnnouncer`, which we can use to populate `ipMapping` on startup and avoid the unnecessary update if a host resolves to an address we've already advertised.